### PR TITLE
Update ValidateCertificateChain() To Allow Certificate Chain Validation without Intermediate Certificate

### LINKED ISSUE #22428: [1.0] Trust validation for CD signing keys missing
### Reproduction steps

```bash
1. Try to add CD verifying keys with `--cd-trust-store-path PATH_TO_CERTS` on chip-tool command line
2. Verify attestation with a CD signed by a key that has its public cert in `PATH_TO_CERTS`, but does not chain to the CSA root key
3. Observe that attestation succeeds, even though the key should not have been trusted and attestation should have failed.

Only signing key certs chaining back to the approved CSA root key (see below) must be possible to add to the `CsaCdKeysTrustStore`.


-----BEGIN CERTIFICATE-----
MIICATCCAaegAwIBAgIHY3Nhcm9vdDAKBggqhkjOPQQDAjBSMQwwCgYDVQQKDAND
U0ExLDAqBgNVBAMMI01hdHRlciBDZXJ0aWZpY2F0aW9uIGFuZCBUZXN0aW5nIENB
MRQwEgYKKwYBBAGConwCAQwEQzVBMDAgFw0yMjA3MDcxOTI4MDRaGA8yMTIyMDYx
MzE5MjgwNFowUjEMMAoGA1UECgwDQ1NBMSwwKgYDVQQDDCNNYXR0ZXIgQ2VydGlm
aWNhdGlvbiBhbmQgVGVzdGluZyBDQTEUMBIGCisGAQQBgqJ8AgEMBEM1QTAwWTAT
BgcqhkjOPQIBBggqhkjOPQMBBwNCAAQ4SjrDql2+y3IP5iEdPK1IYm/3EaCkkp+t
2GD44nf/wN4fPrYzejSEe1o6BW6ocQ6Td+7t7iUXA/3ZNQEly45Io2YwZDASBgNV
HRMBAf8ECDAGAQH/AgEBMA4GA1UdDwEB/wQEAwIBBjAdBgNVHQ4EFgQUl+Rp0MUE
FMJvxwH3fpR3OQmN9qUwHwYDVR0jBBgwFoAUl+Rp0MUEFMJvxwH3fpR3OQmN9qUw
CgYIKoZIzj0EAwIDSAAwRQIgearlB0fCJ49UoJ6xwKPdlPEopCOL9jVCviODEleI
+mQCIQDvvDCKi7kvj4R4BoFS4BVZGCk4zJ84W4tfTTfu89lRbQ==
-----END CERTIFICATE-----
```
```


### Bug prevalence

100% of the time

### GitHub hash of the SDK that was being used

aa32ccd2433edccfb9ff5b0f22939c2873a30d1b

### Platform

core

### Platform Version(s)

_No response_

### Type

Spec Compliance Issue

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by emargolis**
#### Issue Being Resolved
Fixes #22428 

#### Change overview
Using this function to validate CD signing certificates added to the CD Trust Store.

Added new unit-test cases.

### CODE REVIEW

File: src/crypto/CHIPCryptoPALOpenSSL.cpp [hunk 1]
```diff
@@ -1640,12 +1638,15 @@ CHIP_ERROR ValidateCertificateChain(const uint8_t * rootCertificate, size_t root
     status = X509_STORE_add_cert(store, x509RootCertificate);
     VerifyOrExit(status == 1, (result = CertificateChainValidationResult::kInternalFrameworkError, err = CHIP_ERROR_INTERNAL));
 
-    x509CACertificate = d2i_X509(nullptr, &caCertificate, static_cast<long>(caCertificateLen));
-    VerifyOrExit(x509CACertificate != nullptr,
-                 (result = CertificateChainValidationResult::kICAFormatInvalid, err = CHIP_ERROR_INTERNAL));
+    if (caCertificate != nullptr && caCertificateLen > 0)
+    {
+        x509CACertificate = d2i_X509(nullptr, &caCertificate, static_cast<long>(caCertificateLen));
```
Comment (bzbarsky-apple) [inline src/crypto/CHIPCryptoPALOpenSSL.cpp:1643]: This lost the CanCastTo check, right?

**Review by bzbarsky-apple — APPROVED** [review]

File: src/crypto/CHIPCryptoPALOpenSSL.cpp (hunk 1 above)
Comment (emargolis) [reply src/crypto/CHIPCryptoPALOpenSSL.cpp:1647]: Yes, it looks that these checks were missing. I just added them.

File: src/credentials/attestation_verifier/DefaultDeviceAttestationVerifier.cpp [hunk 2]
```diff
@@ -450,15 +498,24 @@ CHIP_ERROR CsaCdKeysTrustStore::AddTrustedKey(const ByteSpan & kid, const Crypto
...
     VerifyOrReturnError(CHIP_NO_ERROR == Crypto::ExtractPubkeyFromX509Cert(derCertBytes, pubKey), CHIP_ERROR_INVALID_ARGUMENT);
+
+    if (!IsCdTestKey(kidSpan))
+    {
+        // Verify cert against CSA CD root of trust.
+        CertificateChainValidationResult chainValidationResult;
+        VerifyOrReturnError(CHIP_NO_ERROR ==
+                                ValidateCertificateChain(gCdRootCert, sizeof(gCdRootCert), nullptr, 0, derCertBytes.data(),
+                                                         derCertBytes.size(), chainValidationResult),
+                            CHIP_ERROR_CERT_NOT_TRUSTED);
```
Comment (tcarmelveilleux) [inline src/credentials/attestation_verifier/DefaultDeviceAttestationVerifier.cpp:515]: The public API states it should give CHIP_ERROR_INVALID_ARGUMENT. Either add this error to the docs in the header, or change this to CHIP_ERROR_INVALID_ARGUMENT.

**Review by tcarmelveilleux — APPROVED** [review]

File: src/credentials/attestation_verifier/DefaultDeviceAttestationVerifier.cpp (hunk 2 above)
Comment (emargolis) [reply src/credentials/attestation_verifier/DefaultDeviceAttestationVerifier.cpp]: replaced with CHIP_ERROR_INVALID_ARGUMENT

**Comment by andy31415** [comment]
Accepted for v1: necessary for certificate chain validation

**Pull request merged**